{"cells":[{"source":"# Change directory to VSCode workspace root so that relative path loads work correctly. Turn this addition off with the DataScience.changeDirOnImportExport setting\r\nimport os\r\ntry:\r\n\tos.chdir(os.path.join(os.getcwd(), '..\\..'))\r\n\tprint(os.getcwd())\r\nexcept:\r\n\tpass\r\n","cell_type":"code","outputs":[],"metadata":{},"execution_count":0},{"cell_type":"code","execution_count":1,"metadata":{},"outputs":[],"source":["from stringtieTools.read_files import read_genes_exp, read_transcripts_exp\n","from stringtieTools.gfftools import ncbi_gene_id_lst, ncbi_rna_id_lst, ncbi_add_generalID_to_maskID\n","from stringtieTools.mergeSamples import merge_gene_exp, merge_tx_exp"]},{"cell_type":"code","execution_count":2,"metadata":{},"outputs":[{"data":{"text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>Gene ID</th>\n      <th>Gene Name</th>\n      <th>Coverage</th>\n      <th>FPKM</th>\n      <th>TPM</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>gene0</td>\n      <td>LOC425783</td>\n      <td>61.715477</td>\n      <td>15.702685</td>\n      <td>38.123001</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>gene1</td>\n      <td>LOC107054287</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>gene2</td>\n      <td>LOC107053477</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>gene3</td>\n      <td>LOC430443</td>\n      <td>1.294543</td>\n      <td>0.329379</td>\n      <td>0.799668</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>gene4</td>\n      <td>LOC107053049</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>gene5</td>\n      <td>GOLGB1</td>\n      <td>70.094582</td>\n      <td>17.861826</td>\n      <td>43.364967</td>\n    </tr>\n    <tr>\n      <th>6</th>\n      <td>gene6</td>\n      <td>HCLS1</td>\n      <td>0.229274</td>\n      <td>0.058404</td>\n      <td>0.141793</td>\n    </tr>\n    <tr>\n      <th>7</th>\n      <td>gene7</td>\n      <td>OR9Q1</td>\n      <td>0.148075</td>\n      <td>0.037676</td>\n      <td>0.091469</td>\n    </tr>\n    <tr>\n      <th>8</th>\n      <td>gene8</td>\n      <td>NME6</td>\n      <td>15.499718</td>\n      <td>3.943698</td>\n      <td>9.574515</td>\n    </tr>\n    <tr>\n      <th>9</th>\n      <td>gene9</td>\n      <td>RABL2B</td>\n      <td>13.237181</td>\n      <td>3.368025</td>\n      <td>8.176896</td>\n    </tr>\n    <tr>\n      <th>10</th>\n      <td>gene10</td>\n      <td>ACR</td>\n      <td>0.204806</td>\n      <td>0.052110</td>\n      <td>0.126513</td>\n    </tr>\n    <tr>\n      <th>11</th>\n      <td>gene11</td>\n      <td>LOC107053643</td>\n      <td>40.479965</td>\n      <td>10.299591</td>\n      <td>25.005360</td>\n    </tr>\n    <tr>\n      <th>12</th>\n      <td>gene12</td>\n      <td>LOC107054191</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>13</th>\n      <td>gene13</td>\n      <td>LOC107053478</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>14</th>\n      <td>gene14</td>\n      <td>ARSA</td>\n      <td>74.368622</td>\n      <td>22.443634</td>\n      <td>54.488686</td>\n    </tr>\n    <tr>\n      <th>15</th>\n      <td>gene15</td>\n      <td>LOC769176</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>16</th>\n      <td>gene16</td>\n      <td>SHANK3</td>\n      <td>0.925357</td>\n      <td>0.235445</td>\n      <td>0.571613</td>\n    </tr>\n    <tr>\n      <th>17</th>\n      <td>gene17</td>\n      <td>MIR6708</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>18</th>\n      <td>gene18</td>\n      <td>MIR1604</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>19</th>\n      <td>gene19</td>\n      <td>LOC101749001</td>\n      <td>4.830187</td>\n      <td>1.228977</td>\n      <td>2.983712</td>\n    </tr>\n    <tr>\n      <th>20</th>\n      <td>gene20</td>\n      <td>PAX4</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>21</th>\n      <td>gene21</td>\n      <td>LOC107055071</td>\n      <td>22.628414</td>\n      <td>5.757500</td>\n      <td>13.978066</td>\n    </tr>\n    <tr>\n      <th>22</th>\n      <td>gene22</td>\n      <td>GCC1</td>\n      <td>48.493374</td>\n      <td>12.338496</td>\n      <td>29.955418</td>\n    </tr>\n    <tr>\n      <th>23</th>\n      <td>gene23</td>\n      <td>DENND6B</td>\n      <td>33.068115</td>\n      <td>10.776205</td>\n      <td>26.162487</td>\n    </tr>\n    <tr>\n      <th>24</th>\n      <td>gene24</td>\n      <td>PPP6R2</td>\n      <td>9.362923</td>\n      <td>14.045566</td>\n      <td>34.099846</td>\n    </tr>\n    <tr>\n      <th>25</th>\n      <td>gene25</td>\n      <td>MIR1651</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>26</th>\n      <td>gene26</td>\n      <td>SBF1</td>\n      <td>117.090439</td>\n      <td>30.534409</td>\n      <td>74.131477</td>\n    </tr>\n    <tr>\n      <th>27</th>\n      <td>gene27</td>\n      <td>MIOX</td>\n      <td>0.432103</td>\n      <td>0.111721</td>\n      <td>0.271236</td>\n    </tr>\n    <tr>\n      <th>28</th>\n      <td>gene28</td>\n      <td>LMF2</td>\n      <td>320.624512</td>\n      <td>86.140541</td>\n      <td>209.132126</td>\n    </tr>\n    <tr>\n      <th>29</th>\n      <td>gene29</td>\n      <td>NCAPH2</td>\n      <td>37.266811</td>\n      <td>13.812909</td>\n      <td>33.535000</td>\n    </tr>\n    <tr>\n      <th>...</th>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n    </tr>\n    <tr>\n      <th>26590</th>\n      <td>gene26610</td>\n      <td>USP39</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>26591</th>\n      <td>gene26611</td>\n      <td>SIGLEC1</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>26592</th>\n      <td>gene26612</td>\n      <td>LOC107051568</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>26593</th>\n      <td>gene26613</td>\n      <td>LOC430766</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>26594</th>\n      <td>gene26614</td>\n      <td>LOC107051570</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>26595</th>\n      <td>gene26615</td>\n      <td>LOC101750306</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>26596</th>\n      <td>gene26616</td>\n      <td>LOC101751171</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>26597</th>\n      <td>gene26617</td>\n      <td>LOC107051571</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>26598</th>\n      <td>gene26618</td>\n      <td>LOC426220</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>26599</th>\n      <td>gene26619</td>\n      <td>LOC101750469</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>26600</th>\n      <td>gene26620</td>\n      <td>LOC101749414</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>26601</th>\n      <td>gene26621</td>\n      <td>LOC101748465</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>26602</th>\n      <td>gene26622</td>\n      <td>LOC101751341</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>26603</th>\n      <td>gene26623</td>\n      <td>LOC101748190</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>26604</th>\n      <td>gene26624</td>\n      <td>RP11-298I3.5</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>26605</th>\n      <td>gene26625</td>\n      <td>LOC107051572</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>26606</th>\n      <td>gene26626</td>\n      <td>LOC100858934</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>26607</th>\n      <td>gene26634</td>\n      <td>ND3</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>26608</th>\n      <td>gene26639</td>\n      <td>ND6</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>26609</th>\n      <td>gene26628</td>\n      <td>ND2</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>26610</th>\n      <td>gene26632</td>\n      <td>ATP6</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>26611</th>\n      <td>gene26637</td>\n      <td>ND5</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>26612</th>\n      <td>gene26630</td>\n      <td>COX2</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>26613</th>\n      <td>gene26635</td>\n      <td>ND4L</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>26614</th>\n      <td>gene26629</td>\n      <td>COX1</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>26615</th>\n      <td>gene26633</td>\n      <td>COX3</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>26616</th>\n      <td>gene26638</td>\n      <td>CYTB</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>26617</th>\n      <td>gene26627</td>\n      <td>ND1</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>26618</th>\n      <td>gene26631</td>\n      <td>ATP8</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>26619</th>\n      <td>gene26636</td>\n      <td>ND4</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n  </tbody>\n</table>\n<p>26620 rows Ã— 5 columns</p>\n</div>","text/plain":"         Gene ID     Gene Name    Coverage       FPKM         TPM\n0          gene0     LOC425783   61.715477  15.702685   38.123001\n1          gene1  LOC107054287    0.000000   0.000000    0.000000\n2          gene2  LOC107053477    0.000000   0.000000    0.000000\n3          gene3     LOC430443    1.294543   0.329379    0.799668\n4          gene4  LOC107053049    0.000000   0.000000    0.000000\n5          gene5        GOLGB1   70.094582  17.861826   43.364967\n6          gene6         HCLS1    0.229274   0.058404    0.141793\n7          gene7         OR9Q1    0.148075   0.037676    0.091469\n8          gene8          NME6   15.499718   3.943698    9.574515\n9          gene9        RABL2B   13.237181   3.368025    8.176896\n10        gene10           ACR    0.204806   0.052110    0.126513\n11        gene11  LOC107053643   40.479965  10.299591   25.005360\n12        gene12  LOC107054191    0.000000   0.000000    0.000000\n13        gene13  LOC107053478    0.000000   0.000000    0.000000\n14        gene14          ARSA   74.368622  22.443634   54.488686\n15        gene15     LOC769176    0.000000   0.000000    0.000000\n16        gene16        SHANK3    0.925357   0.235445    0.571613\n17        gene17       MIR6708    0.000000   0.000000    0.000000\n18        gene18       MIR1604    0.000000   0.000000    0.000000\n19        gene19  LOC101749001    4.830187   1.228977    2.983712\n20        gene20          PAX4    0.000000   0.000000    0.000000\n21        gene21  LOC107055071   22.628414   5.757500   13.978066\n22        gene22          GCC1   48.493374  12.338496   29.955418\n23        gene23       DENND6B   33.068115  10.776205   26.162487\n24        gene24        PPP6R2    9.362923  14.045566   34.099846\n25        gene25       MIR1651    0.000000   0.000000    0.000000\n26        gene26          SBF1  117.090439  30.534409   74.131477\n27        gene27          MIOX    0.432103   0.111721    0.271236\n28        gene28          LMF2  320.624512  86.140541  209.132126\n29        gene29        NCAPH2   37.266811  13.812909   33.535000\n...          ...           ...         ...        ...         ...\n26590  gene26610         USP39    0.000000   0.000000    0.000000\n26591  gene26611       SIGLEC1    0.000000   0.000000    0.000000\n26592  gene26612  LOC107051568    0.000000   0.000000    0.000000\n26593  gene26613     LOC430766    0.000000   0.000000    0.000000\n26594  gene26614  LOC107051570    0.000000   0.000000    0.000000\n26595  gene26615  LOC101750306    0.000000   0.000000    0.000000\n26596  gene26616  LOC101751171    0.000000   0.000000    0.000000\n26597  gene26617  LOC107051571    0.000000   0.000000    0.000000\n26598  gene26618     LOC426220    0.000000   0.000000    0.000000\n26599  gene26619  LOC101750469    0.000000   0.000000    0.000000\n26600  gene26620  LOC101749414    0.000000   0.000000    0.000000\n26601  gene26621  LOC101748465    0.000000   0.000000    0.000000\n26602  gene26622  LOC101751341    0.000000   0.000000    0.000000\n26603  gene26623  LOC101748190    0.000000   0.000000    0.000000\n26604  gene26624  RP11-298I3.5    0.000000   0.000000    0.000000\n26605  gene26625  LOC107051572    0.000000   0.000000    0.000000\n26606  gene26626  LOC100858934    0.000000   0.000000    0.000000\n26607  gene26634           ND3    0.000000   0.000000    0.000000\n26608  gene26639           ND6    0.000000   0.000000    0.000000\n26609  gene26628           ND2    0.000000   0.000000    0.000000\n26610  gene26632          ATP6    0.000000   0.000000    0.000000\n26611  gene26637           ND5    0.000000   0.000000    0.000000\n26612  gene26630          COX2    0.000000   0.000000    0.000000\n26613  gene26635          ND4L    0.000000   0.000000    0.000000\n26614  gene26629          COX1    0.000000   0.000000    0.000000\n26615  gene26633          COX3    0.000000   0.000000    0.000000\n26616  gene26638          CYTB    0.000000   0.000000    0.000000\n26617  gene26627           ND1    0.000000   0.000000    0.000000\n26618  gene26631          ATP8    0.000000   0.000000    0.000000\n26619  gene26636           ND4    0.000000   0.000000    0.000000\n\n[26620 rows x 5 columns]"},"execution_count":2,"metadata":{},"output_type":"execute_result"}],"source":["read_genes_exp('./stringtieTools/example/Chicken_I089-T05_good_1.abund.txt')\n",""]},{"cell_type":"code","execution_count":3,"metadata":{},"outputs":[{"ename":"TypeError","evalue":"list() takes at most 1 argument (6 given)","output_type":"error","traceback":["\u001b[1;31m---------------------------------------------------------------------------\u001b[0m","\u001b[1;31mTypeError\u001b[0m                                 Traceback (most recent call last)","\u001b[1;32m<ipython-input-3-bf1e473e3b3f>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mread_transcripts_exp\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'stringtieTools\\example\\Chicken_I089-T05_good_1.abund.gtf'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[1;32me:\\R135_remote-ftp\\stringtieTools\\read_files.py\u001b[0m in \u001b[0;36mread_transcripts_exp\u001b[1;34m(ifile)\u001b[0m\n\u001b[0;32m     60\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     61\u001b[0m     features = idf[8].map(\n\u001b[1;32m---> 62\u001b[1;33m         \u001b[0mget_values\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     63\u001b[0m     \u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mstr\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     64\u001b[0m         \u001b[1;34m';'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mexpand\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mTrue\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n","\u001b[1;32mG:\\install_place\\anaconda3\\lib\\site-packages\\pandas\\core\\series.py\u001b[0m in \u001b[0;36mmap\u001b[1;34m(self, arg, na_action)\u001b[0m\n\u001b[0;32m   2994\u001b[0m         \"\"\"\n\u001b[0;32m   2995\u001b[0m         new_values = super(Series, self)._map_values(\n\u001b[1;32m-> 2996\u001b[1;33m             arg, na_action=na_action)\n\u001b[0m\u001b[0;32m   2997\u001b[0m         return self._constructor(new_values,\n\u001b[0;32m   2998\u001b[0m                                  index=self.index).__finalize__(self)\n","\u001b[1;32mG:\\install_place\\anaconda3\\lib\\site-packages\\pandas\\core\\base.py\u001b[0m in \u001b[0;36m_map_values\u001b[1;34m(self, mapper, na_action)\u001b[0m\n\u001b[0;32m   1002\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1003\u001b[0m         \u001b[1;31m# mapper is a function\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1004\u001b[1;33m         \u001b[0mnew_values\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mmap_f\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mmapper\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1005\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1006\u001b[0m         \u001b[1;32mreturn\u001b[0m \u001b[0mnew_values\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n","\u001b[1;32mpandas/_libs/src\\inference.pyx\u001b[0m in \u001b[0;36mpandas._libs.lib.map_infer\u001b[1;34m()\u001b[0m\n","\u001b[1;32me:\\R135_remote-ftp\\stringtieTools\\read_files.py\u001b[0m in \u001b[0;36mget_values\u001b[1;34m(x)\u001b[0m\n\u001b[0;32m     55\u001b[0m                 \u001b[0mre_search\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34mr'cov \"([^\\\"]+)\"'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mx\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     56\u001b[0m                 \u001b[0mre_search\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34mr'FPKM \"([^\\\"]+)\"'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mx\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 57\u001b[1;33m                 \u001b[0mre_search\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34mr'TPM \"([^\\\"]+)\"'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mx\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     58\u001b[0m             )\n\u001b[0;32m     59\u001b[0m         )\n","\u001b[1;31mTypeError\u001b[0m: list() takes at most 1 argument (6 given)"]}],"source":["read_transcripts_exp('stringtieTools\\example\\Chicken_I089-T05_good_1.abund.gtf')\n",""]},{"cell_type":"code","execution_count":4,"metadata":{},"outputs":[{"data":{"text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>mask_id</th>\n      <th>general_id</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>gene0</td>\n      <td>425783</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>gene1</td>\n      <td>107054287</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>gene2</td>\n      <td>107053477</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>gene3</td>\n      <td>430443</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>gene4</td>\n      <td>107053049</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>gene5</td>\n      <td>426868</td>\n    </tr>\n  </tbody>\n</table>\n</div>","text/plain":"  mask_id general_id\n0   gene0     425783\n1   gene1  107054287\n2   gene2  107053477\n3   gene3     430443\n4   gene4  107053049\n5   gene5     426868"},"execution_count":4,"metadata":{},"output_type":"execute_result"}],"source":["ncbi_gene_id_lst('stringtieTools\\example\\gal5_ncbi.refseq.head100.gff')\n",""]},{"cell_type":"code","execution_count":5,"metadata":{},"outputs":[{"data":{"text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>mask_id</th>\n      <th>general_id</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>rna0</td>\n      <td>XM_015282442.1</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>rna1</td>\n      <td>XR_001469189.1</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>rna2</td>\n      <td>XR_001467601.1</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>rna3</td>\n      <td>XM_428001.5</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>rna4</td>\n      <td>XM_015284588.1</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>rna5</td>\n      <td>XM_015290459.1</td>\n    </tr>\n    <tr>\n      <th>6</th>\n      <td>rna6</td>\n      <td>XM_015291512.1</td>\n    </tr>\n  </tbody>\n</table>\n</div>","text/plain":"  mask_id      general_id\n0    rna0  XM_015282442.1\n1    rna1  XR_001469189.1\n2    rna2  XR_001467601.1\n3    rna3     XM_428001.5\n4    rna4  XM_015284588.1\n5    rna5  XM_015290459.1\n6    rna6  XM_015291512.1"},"execution_count":5,"metadata":{},"output_type":"execute_result"}],"source":["ncbi_rna_id_lst('stringtieTools\\example\\gal5_ncbi.refseq.head100.gff')\n",""]},{"cell_type":"code","execution_count":6,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":"<Input DataFrame is:>\n   Gene ID     Gene Name   Coverage       FPKM        TPM\n0   gene0     LOC425783  61.715477  15.702685  38.123001\n1   gene1  LOC107054287   0.000000   0.000000   0.000000\n2   gene2  LOC107053477   0.000000   0.000000   0.000000\n3   gene3     LOC430443   1.294543   0.329379   0.799668\n4   gene4  LOC107053049   0.000000   0.000000   0.000000\n<Making list of RNA IDs and GeneBank IDs>\n<Merge DataFrames by column \"Gene ID\">\n<Input 26620 rows and output 6 rows, since some Gene refseq IDs have no Gene IDs.>\n"},{"data":{"text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>Gene ID</th>\n      <th>general_id</th>\n      <th>Gene Name</th>\n      <th>Coverage</th>\n      <th>FPKM</th>\n      <th>TPM</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>gene0</td>\n      <td>425783</td>\n      <td>LOC425783</td>\n      <td>61.715477</td>\n      <td>15.702685</td>\n      <td>38.123001</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>gene1</td>\n      <td>107054287</td>\n      <td>LOC107054287</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>gene2</td>\n      <td>107053477</td>\n      <td>LOC107053477</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>gene3</td>\n      <td>430443</td>\n      <td>LOC430443</td>\n      <td>1.294543</td>\n      <td>0.329379</td>\n      <td>0.799668</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>gene4</td>\n      <td>107053049</td>\n      <td>LOC107053049</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>gene5</td>\n      <td>426868</td>\n      <td>GOLGB1</td>\n      <td>70.094582</td>\n      <td>17.861826</td>\n      <td>43.364967</td>\n    </tr>\n  </tbody>\n</table>\n</div>","text/plain":"  Gene ID general_id     Gene Name   Coverage       FPKM        TPM\n0   gene0     425783     LOC425783  61.715477  15.702685  38.123001\n1   gene1  107054287  LOC107054287   0.000000   0.000000   0.000000\n2   gene2  107053477  LOC107053477   0.000000   0.000000   0.000000\n3   gene3     430443     LOC430443   1.294543   0.329379   0.799668\n4   gene4  107053049  LOC107053049   0.000000   0.000000   0.000000\n5   gene5     426868        GOLGB1  70.094582  17.861826  43.364967"},"execution_count":6,"metadata":{},"output_type":"execute_result"}],"source":["ncbi_add_generalID_to_maskID(\n","    'gene', \n","    read_genes_exp(\n","    './stringtieTools/example/Chicken_I089-T05_good_1.abund.txt'),\n","    'Gene ID',\n","    'stringtieTools\\example\\gal5_ncbi.refseq.head100.gff'\n",")\n","\n",""]},{"cell_type":"code","execution_count":7,"metadata":{},"outputs":[{"ename":"TypeError","evalue":"list() takes at most 1 argument (6 given)","output_type":"error","traceback":["\u001b[1;31m---------------------------------------------------------------------------\u001b[0m","\u001b[1;31mTypeError\u001b[0m                                 Traceback (most recent call last)","\u001b[1;32m<ipython-input-7-e4fca99b046d>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      2\u001b[0m     \u001b[1;34m'tx'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      3\u001b[0m     read_transcripts_exp(\n\u001b[1;32m----> 4\u001b[1;33m         'stringtieTools\\example\\Chicken_I089-T05_good_1.abund.gtf'),\n\u001b[0m\u001b[0;32m      5\u001b[0m     \u001b[1;34m'transcript_id'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      6\u001b[0m     \u001b[1;34m'stringtieTools\\example\\gal5_ncbi.refseq.head100.gff'\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n","\u001b[1;32me:\\R135_remote-ftp\\stringtieTools\\read_files.py\u001b[0m in \u001b[0;36mread_transcripts_exp\u001b[1;34m(ifile)\u001b[0m\n\u001b[0;32m     60\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     61\u001b[0m     features = idf[8].map(\n\u001b[1;32m---> 62\u001b[1;33m         \u001b[0mget_values\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     63\u001b[0m     \u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mstr\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     64\u001b[0m         \u001b[1;34m';'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mexpand\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mTrue\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n","\u001b[1;32mG:\\install_place\\anaconda3\\lib\\site-packages\\pandas\\core\\series.py\u001b[0m in \u001b[0;36mmap\u001b[1;34m(self, arg, na_action)\u001b[0m\n\u001b[0;32m   2994\u001b[0m         \"\"\"\n\u001b[0;32m   2995\u001b[0m         new_values = super(Series, self)._map_values(\n\u001b[1;32m-> 2996\u001b[1;33m             arg, na_action=na_action)\n\u001b[0m\u001b[0;32m   2997\u001b[0m         return self._constructor(new_values,\n\u001b[0;32m   2998\u001b[0m                                  index=self.index).__finalize__(self)\n","\u001b[1;32mG:\\install_place\\anaconda3\\lib\\site-packages\\pandas\\core\\base.py\u001b[0m in \u001b[0;36m_map_values\u001b[1;34m(self, mapper, na_action)\u001b[0m\n\u001b[0;32m   1002\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1003\u001b[0m         \u001b[1;31m# mapper is a function\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1004\u001b[1;33m         \u001b[0mnew_values\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mmap_f\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mmapper\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1005\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1006\u001b[0m         \u001b[1;32mreturn\u001b[0m \u001b[0mnew_values\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n","\u001b[1;32mpandas/_libs/src\\inference.pyx\u001b[0m in \u001b[0;36mpandas._libs.lib.map_infer\u001b[1;34m()\u001b[0m\n","\u001b[1;32me:\\R135_remote-ftp\\stringtieTools\\read_files.py\u001b[0m in \u001b[0;36mget_values\u001b[1;34m(x)\u001b[0m\n\u001b[0;32m     55\u001b[0m                 \u001b[0mre_search\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34mr'cov \"([^\\\"]+)\"'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mx\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     56\u001b[0m                 \u001b[0mre_search\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34mr'FPKM \"([^\\\"]+)\"'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mx\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 57\u001b[1;33m                 \u001b[0mre_search\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34mr'TPM \"([^\\\"]+)\"'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mx\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     58\u001b[0m             )\n\u001b[0;32m     59\u001b[0m         )\n","\u001b[1;31mTypeError\u001b[0m: list() takes at most 1 argument (6 given)"]}],"source":["ncbi_add_generalID_to_maskID(\n","    'tx',\n","    read_transcripts_exp(\n","        'stringtieTools\\example\\Chicken_I089-T05_good_1.abund.gtf'),\n","    'transcript_id',\n","    'stringtieTools\\example\\gal5_ncbi.refseq.head100.gff'\n",")\n","\n",""]},{"cell_type":"code","execution_count":8,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":"<Get expression in \"FPKM\" method.>\n<Reading :> Chicken_I089-T05_good_1.abund.txt\n<Input DataFrame is:>\n   Gene ID     Gene Name  Chicken_I089-T05_good_1\n0   gene0     LOC425783                15.702685\n1   gene1  LOC107054287                 0.000000\n2   gene2  LOC107053477                 0.000000\n3   gene3     LOC430443                 0.329379\n4   gene4  LOC107053049                 0.000000\n<Making list of RNA IDs and GeneBank IDs>\n<Merge DataFrames by column \"Gene ID\">\n<Input 26620 rows and output 6 rows, since some Gene refseq IDs have no Gene IDs.>\n"},{"data":{"text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>Gene ID</th>\n      <th>general_id</th>\n      <th>Gene Name</th>\n      <th>Chicken_I089-T05_good_1</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>gene0</td>\n      <td>425783</td>\n      <td>LOC425783</td>\n      <td>15.702685</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>gene1</td>\n      <td>107054287</td>\n      <td>LOC107054287</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>gene2</td>\n      <td>107053477</td>\n      <td>LOC107053477</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>gene3</td>\n      <td>430443</td>\n      <td>LOC430443</td>\n      <td>0.329379</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>gene4</td>\n      <td>107053049</td>\n      <td>LOC107053049</td>\n      <td>0.000000</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>gene5</td>\n      <td>426868</td>\n      <td>GOLGB1</td>\n      <td>17.861826</td>\n    </tr>\n  </tbody>\n</table>\n</div>","text/plain":"  Gene ID general_id     Gene Name  Chicken_I089-T05_good_1\n0   gene0     425783     LOC425783                15.702685\n1   gene1  107054287  LOC107054287                 0.000000\n2   gene2  107053477  LOC107053477                 0.000000\n3   gene3     430443     LOC430443                 0.329379\n4   gene4  107053049  LOC107053049                 0.000000\n5   gene5     426868        GOLGB1                17.861826"},"execution_count":8,"metadata":{},"output_type":"execute_result"}],"source":["merge_gene_exp(\n","    'stringtieTools\\example', 'stringtieTools\\example\\gal5_ncbi.refseq.head100.gff',\n","    exp_type='fpkm',\n",")\n","\n",""]},{"cell_type":"code","execution_count":9,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":"<Get expression in \"FPKM\" method.>\n<Reading :> Chicken_I089-T05_good_1.abund.gtf\n"},{"ename":"TypeError","evalue":"list() takes at most 1 argument (6 given)","output_type":"error","traceback":["\u001b[1;31m---------------------------------------------------------------------------\u001b[0m","\u001b[1;31mTypeError\u001b[0m                                 Traceback (most recent call last)","\u001b[1;32m<ipython-input-9-81443e9d0197>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      1\u001b[0m merge_tx_exp(\n\u001b[0;32m      2\u001b[0m     \u001b[1;34m'stringtieTools\\example'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'stringtieTools\\example\\gal5_ncbi.refseq.head100.gff'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 3\u001b[1;33m     \u001b[0mexp_type\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'fpkm'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      4\u001b[0m )\n","\u001b[1;32me:\\R135_remote-ftp\\stringtieTools\\mergeSamples.py\u001b[0m in \u001b[0;36mmerge_tx_exp\u001b[1;34m(idir, igff, exp_type, ofile, if_convert_id)\u001b[0m\n\u001b[0;32m     45\u001b[0m             \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'<Reading :>'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mf\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     46\u001b[0m             \u001b[0mifile\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mos\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0midir\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mf\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 47\u001b[1;33m             tmp_df = read_transcripts_exp(ifile)[\n\u001b[0m\u001b[0;32m     48\u001b[0m                 ['gene_id',\t'transcript_id', \n\u001b[0;32m     49\u001b[0m                 'ref_gene_name', exp_type]\n","\u001b[1;32me:\\R135_remote-ftp\\stringtieTools\\read_files.py\u001b[0m in \u001b[0;36mread_transcripts_exp\u001b[1;34m(ifile)\u001b[0m\n\u001b[0;32m     60\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     61\u001b[0m     features = idf[8].map(\n\u001b[1;32m---> 62\u001b[1;33m         \u001b[0mget_values\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     63\u001b[0m     \u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mstr\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     64\u001b[0m         \u001b[1;34m';'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mexpand\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mTrue\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n","\u001b[1;32mG:\\install_place\\anaconda3\\lib\\site-packages\\pandas\\core\\series.py\u001b[0m in \u001b[0;36mmap\u001b[1;34m(self, arg, na_action)\u001b[0m\n\u001b[0;32m   2994\u001b[0m         \"\"\"\n\u001b[0;32m   2995\u001b[0m         new_values = super(Series, self)._map_values(\n\u001b[1;32m-> 2996\u001b[1;33m             arg, na_action=na_action)\n\u001b[0m\u001b[0;32m   2997\u001b[0m         return self._constructor(new_values,\n\u001b[0;32m   2998\u001b[0m                                  index=self.index).__finalize__(self)\n","\u001b[1;32mG:\\install_place\\anaconda3\\lib\\site-packages\\pandas\\core\\base.py\u001b[0m in \u001b[0;36m_map_values\u001b[1;34m(self, mapper, na_action)\u001b[0m\n\u001b[0;32m   1002\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1003\u001b[0m         \u001b[1;31m# mapper is a function\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1004\u001b[1;33m         \u001b[0mnew_values\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mmap_f\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mmapper\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1005\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1006\u001b[0m         \u001b[1;32mreturn\u001b[0m \u001b[0mnew_values\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n","\u001b[1;32mpandas/_libs/src\\inference.pyx\u001b[0m in \u001b[0;36mpandas._libs.lib.map_infer\u001b[1;34m()\u001b[0m\n","\u001b[1;32me:\\R135_remote-ftp\\stringtieTools\\read_files.py\u001b[0m in \u001b[0;36mget_values\u001b[1;34m(x)\u001b[0m\n\u001b[0;32m     55\u001b[0m                 \u001b[0mre_search\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34mr'cov \"([^\\\"]+)\"'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mx\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     56\u001b[0m                 \u001b[0mre_search\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34mr'FPKM \"([^\\\"]+)\"'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mx\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 57\u001b[1;33m                 \u001b[0mre_search\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34mr'TPM \"([^\\\"]+)\"'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mx\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     58\u001b[0m             )\n\u001b[0;32m     59\u001b[0m         )\n","\u001b[1;31mTypeError\u001b[0m: list() takes at most 1 argument (6 given)"]}],"source":["merge_tx_exp(\n","    'stringtieTools\\example', 'stringtieTools\\example\\gal5_ncbi.refseq.head100.gff',\n","    exp_type='fpkm',\n",")\n","\n",""]},{"cell_type":"code","execution_count":10,"metadata":{},"outputs":[],"source":[""]},{"cell_type":"code","execution_count":11,"metadata":{},"outputs":[],"source":""}],"nbformat":4,"nbformat_minor":2,"metadata":{"language_info":{"name":"python","codemirror_mode":{"name":"ipython","version":3}},"orig_nbformat":2,"file_extension":".py","mimetype":"text/x-python","name":"python","npconvert_exporter":"python","pygments_lexer":"ipython3","version":3}}